jetty-websocket/websocket-javax-common/src/main/java/org/eclipse/jetty/websocket/javax/common/messages/DecodedTextMessageSink.java

```diff
@@ -20,61 +20,58 @@
 import org.eclipse.jetty.websocket.javax.common.JavaxWebSocketFrameHandlerFactory;
+import org.eclipse.jetty.websocket.javax.common.decoders.RegisteredDecoder;
 import org.eclipse.jetty.websocket.util.messages.MessageSink;
 import org.eclipse.jetty.websocket.util.messages.StringMessageSink;
 
-public class DecodedTextMessageSink<T> extends DecodedMessageSink<Decoder.Text<T>>
+public class DecodedTextMessageSink<T> extends AbstractDecodedMessageSink.Basic<Decoder.Text<T>>
 {
-    public DecodedTextMessageSink(CoreSession session,
-                                  Decoder.Text<T> decoder,
-                                  MethodHandle methodHandle)
-        throws NoSuchMethodException, IllegalAccessException
+    public DecodedTextMessageSink(CoreSession session, MethodHandle methodHandle, List<RegisteredDecoder> decoders)
     {
-        super(session, decoder, methodHandle);
+        super(session, methodHandle, decoders);
     }
 
     @Override
-    protected MethodHandle newRawMethodHandle() throws NoSuchMethodException, IllegalAccessException
+    MessageSink getMessageSink() throws NoSuchMethodException, IllegalAccessException
     {
-        return JavaxWebSocketFrameHandlerFactory.getServerMethodHandleLookup().findVirtual(DecodedTextMessageSink.class,
-            "onWholeMessage", MethodType.methodType(void.class, String.class))
+        MethodHandle methodHandle = JavaxWebSocketFrameHandlerFactory.getServerMethodHandleLookup()
+            .findVirtual(getClass(), "onMessage", MethodType.methodType(void.class, String.class))
             .bindTo(this);
-    }
-
-    @Override
-    protected MessageSink newRawMessageSink(CoreSession session, MethodHandle rawMethodHandle)
-    {
-        return new StringMessageSink(session, rawMethodHandle);
+        return new StringMessageSink(_coreSession, methodHandle);
     }
 
     @SuppressWarnings("Duplicates")
-    public void onWholeMessage(String wholeMessage)
+    public void onMessage(String wholeMessage)
     {
-        if (!getDecoder().willDecode(wholeMessage))
+        for (Decoder.Text<T> decoder : _decoders)
         {
-            logger.warn("Message lost, decoder " + getDecoder().getClass().getName() + "#willDecode() has rejected it.");
-            return;
+            if (decoder.willDecode(wholeMessage))
+            {
+                try
+                {
+                    T obj = decoder.decode(wholeMessage);
+                    _methodHandle.invoke(obj);
+                    return;
+                }
+                catch (DecodeException e)
+                {
+                    throw new CloseException(CloseReason.CloseCodes.CANNOT_ACCEPT.getCode(), "Unable to decode", e);
+                }
+                catch (Throwable t)
+                {
+                    throw new CloseException(CloseReason.CloseCodes.CANNOT_ACCEPT.getCode(), "Endpoint notification error", t);
+                }
+            }
```

Comment:
This block is duplicated for the binary version. Can it be factored out into the base class?

Reply:
I tried factoring this out to the base class but couldn't do this well because there is no common base interface (something like `Decoder.Whole`) which has the `willDecode()` method, all the types inherit directly from `Decoder`. So I couldn't work out a way to remove this duplication.